include/oxen/quic/loop.hpp

```diff
@@ -124,7 +183,75 @@ namespace oxen::quic
             return fut.get();
         }
 
-        void call_soon(std::function<void(void)> f);
+        /** This overload of `call_every` will begin an indefinitely repeating object tied to the lifetime of `caller`.
+            Prior to executing each iteration, the weak_ptr will be checked to ensure the calling object lifetime has
+            persisted up to that point.
+        */
+        template <typename Callable>
+        void call_every(loop_time interval, std::weak_ptr<void> caller, Callable&& f)
+        {
+            auto handler = make_handler();
+            // grab the reference before giving ownership of the repeater to the lambda
+            auto& h = *handler;
+
+            h.start_event(
+                    loop(),
+                    interval,
+                    [hndlr = std::move(handler), owner = std::move(caller), func = std::forward<Callable>(f)]() mutable {
+                        if (auto ptr = owner.lock())
+                            func();
+                        else
+                            hndlr.reset();
+                    });
+        }
+
+        /** This overload of `call_every` will return an EventHandler object from which the application can start and stop
+            the repeated event. It is NOT tied to the lifetime of the caller via a weak_ptr. If the application wants
+            to defer start until explicitly calling EventHandler::start(), `start_immediately` should take a false boolean.
+        */
+        template <typename Callable>
+        [[nodiscard]] std::shared_ptr<EventHandler> call_every(
+                loop_time interval, Callable&& f, bool start_immediately = true)
+        {
+            auto h = make_handler();
+
+            h->start_event(loop(), interval, std::forward<Callable>(f), true, start_immediately);
+
+            return h;
+        }
+
+        template <typename Callable>
+        void call_later(loop_time delay, Callable&& hook)
+        {
+            if (in_event_loop())
+            {
+                add_oneshot_event(delay, std::forward<Callable>(hook));
+            }
+            else
+            {
+                call_soon([this,
+                           func = std::forward<Callable>(hook),
+                           target_time = get_timestamp<loop_time>() + delay]() mutable {
+                    auto updated_delay = target_time - get_timestamp<loop_time>();
```

Comment:
I'm using a different `get_timestamp` that takes a template for explicit type safety. Also, the only difference between `get_time` and `get_timestamp` is `.time_since_epoch()`
```c++
std::chrono::steady_clock::time_point get_time()
{
    return std::chrono::steady_clock::now();
}
std::chrono::nanoseconds get_timestamp()
{
    return std::chrono::steady_clock::now().time_since_epoch();
}
```

Reply:
> Also, the only difference between get_time and get_timestamp is .time_since_epoch()

That *is* the important difference.  The former is an abitrary duration that is meaningless on its own; the latter is a object that is a duration with an epoch that fully defines it.

Reply:
Okay so `get_time` is an arbitrary duration that is meaningless on its own? So makes sense why we shouldn't use that one

Reply:
No, other way around.  I had former/latter backwards in my comment.

Reply:
If you can write this out inline and show me it's simpler and an improvement I will happily make the change, but I don't see your argument on this

Reply:
> I think if you write this out you will see that it's effectively the same

Effectively the same in its final value, but different in its semantic meaning.

You can write all sorts of things that are effectively the same, but the use of types (i.e. abstractions) is to give semantic meaning through your intermediate types of what it is that you hold in your hand.  std::chrono::nanosecond is, under the hood, just an integer, but the type adds meaning to that integer.

A std::chrono::time_point is, under the hood, just a duration with an implied reference point, but that still matters, because nanoseconds with a reference point has a different semantic meaning than nanoseconds alone.

Reply:
Except that taking an `std::chrono` type (milliseconds, nanoseconds, etc) and duration casting it is the simpler way to do it. Taking a time point requires either appending `.time_since_epoch()` (which makes it *exactly* like `get_timestamp()`) or finding the difference between it and another time_point, at which point one must duration cast...

Reply:
I don't know why the comment I wrote on this yesterday isn't here, so posting it again:

The principle here is that you don't throw away type safety before you need to without good reason: which means you should keep time points as time point objects, not durations which lose some semantic meaning.

The code here to keep time points as time points is not any more complicated (and only involves one duration cast instead of two), while not throwing away the type safety in intermediate steps:
```C++
                call_soon([this,
                           func = std::move(hook),
                           target_time = get_time() + delay]() mutable {
                    auto now = get_time();

                    if (now >= target_time)
                        func();
                    else
                        add_oneshot_event(
                                std::chrono::duration_cast<std::chrono::microseconds>(target_time - now),
                                std::move(func));
                });
```